compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -103,20 +103,28 @@ impl<'a> AstValidator<'a> {
         self.is_tilde_const_allowed = old;
     }
 
-    fn with_let_allowed(&mut self, allowed: bool, f: impl FnOnce(&mut Self, bool)) {
-        let old = mem::replace(&mut self.is_let_allowed, allowed);
+    fn with_let_management(
+        &mut self,
+        let_management_ty: LetManagementTy,
+        f: impl FnOnce(&mut Self, LetManagementTy),
+    ) {
```

Comment:
might be better as
```suggestion
    fn with_let_forbidden(
        &mut self,
        let_forbidden: Option<LetForbidReason>,
        f: impl FnOnce(&mut Self, Option<LetForbidReason>),
    ) {
```
which should allow the merging of the 2 match arms for `let` below and also
make the second argument to `ban_let_expr` more obvious.